### DISCUSSION TITLE
Practice - Design Splitwise

### DISCUSSION THREAD

**Discussion by kanmaytacker** [kanmaytacker/design-questions#9] [General]
What?

Identify the major entities and their attributes in an expense management system ([Splitwise](https://www.splitwise.com/))
Create a class diagram for the system.

Why?
Machine coding of a management system.

How?
Follow the instructions [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/07-design-splitwise-worksheet.md).

**Comment by Soumikpal9** [comment]
[Splitwise_Entities.txt](https://github.com/kanmaytacker/design-questions/files/9953637/Splitwise_Entities.txt)

**Reply by kanmaytacker** [reply]
@Soumikpal9 You have the correct set of entities. However, some attributes are missing in the Expense class.
Think about who all can pay and similarly how do we identify who all are owed money.

**Comment by akash-bhardwaj** [comment]
User
  - String name
  - String phone
  - String email
  - Long amountOwed
  - Long owesAmount
  - String[] groupNames
  - Expense[] expenses
  
Group
 - User[] users
 
Expense
 - String groupName
 - String paidBy
 - String paidTo
 - Long amount
 - String description

**Reply by kanmaytacker** [reply]
@akash-bhardwaj Similar feedback as above. You have identified the major entities.
The are some minor attributes missing in Group.
However, how do we identify how do we owe money? Similarly, can just one user pay?

**Comment by vkartik** [comment]
class User
{
string name
string username
string email
string password
}

class Expense
{
string Id
double amount
double spentamout
datetime createdat
user paidby
list<user> participants
list<split> splitfortheexpense
}

class split
{
user participant
double percentage
double amount
}

class Group
{
string name
list<user> admin
List<user> members
user createdby
}

**Comment by navnnkk** [comment]
```mermaid
classDiagram

class User {
    - name
    - phone number
    - password
    - group[]
}

class Group {
    - name
    - description
    - user[]
}

class Transaction {
    - id
    - group
    - amount
    - paid by
    - paid to
}


```

**Reply by sastepankaj** [reply]
![splitwise](https://user-images.githubusercontent.com/13418342/200365211-ffe02e95-5e81-4fda-86e5-a3157d341d49.jpg)

**Comment by Faizahmadfaiz** [comment]
`1. User
   - owedAmount
   - toPayAmount
   - Expense[]
   - Group[]
2. Profile
    - name
    - phone
    - password
3. Expense
    - Group
    - Users
    - Transaction
    - description
4. Group
   - admin: User
   - members: User[]
5. Transaction
    - paidBy
    - paidAmount
    - participants: User[]`

**Comment by kk-engineer** [comment]
```mermaid 
classDiagram

    class User {
        -String name
        -String email
        -String password
        -String username
    }
    
    class Expense {
        -String description
        -Double amount
        -Currency currency
        -DateTime createdAt
        -List[User] participants
        -Group group     
    }
    
    class Group {
        -String name
        -List[User] members
        -List[USer] admins
        -User createdBy
        -DateTime cratedAt
    }
    
    class ExpensePaid {
        -User user
        -Expense Expense
    }
    
    class ExpenseOwed {
        -User User
        -Expense Expense
    }
    
    class ExpenseGroup {
        -Expense expense
        -Group group
    }
    
    Expense "1" --o  "*" User
    Expense "1" --o "1" Group
    Group "1" --o "*" User
    
    ExpensePaid "1" --o "1" User
    ExpensePaid "1" --o  "1" Expense
    
    ExpenseOwed "1" --o "1"  User
    ExpenseOwed "1" --o "1"  Expense
    
    ExpenseGroup "1" --o "1" Group
    ExpenseGroup "1" --o "1" Expense

```
